File: services/client_headers.py
"""Client identity defaults and conversation-header forwarding, without credentials."""

import os
from collections.abc import Mapping

CLIENT_HEADER_NAMES = {
    "user-agent": "User-Agent",
    "originator": "Originator",
    "session-id": "Session-Id",
    "thread-id": "Thread-Id",
    "session_id": "session_id",
    "x-session-id": "X-Session-Id",
    "x-codex-session-id": "X-Codex-Session-Id",
    "x-session-affinity": "X-Session-Affinity",
}
OPENCODE_CLIENT_HEADER_NAMES = {
    f"x-opencode-{name}": f"X-Opencode-{name.title()}"
    for name in ("session", "client", "project", "request")
}
SESSION_HEADER_NAMES = (
    "thread-id",
    "session-id",
    "session_id",
    "x-session-id",
    "x-codex-session-id",
    "x-session-affinity",
)


def _value(value: str) -> str:
    if not isinstance(value, str) or any(
        ord(char) < 32 or ord(char) > 126 for char in value
    ):
        return ""
    return value.strip()


def client_context_headers(
    headers: Mapping[str, str], provider: str = ""
) -> dict[str, str]:
    names = {**CLIENT_HEADER_NAMES}
    if provider == "opencode":
        names.update(OPENCODE_CLIENT_HEADER_NAMES)
    return {
        names[name.lower()]: value
        for name, value in headers.items()
        if name.lower() in names and _value(value)
    }


def with_client_defaults(
    headers: Mapping[str, str], provider: str = ""
) -> dict[str, str]:
    """Fill missing identity only; keep explicit identity, auth and protocol headers."""
    result = dict(headers)
    names = {**CLIENT_HEADER_NAMES, **OPENCODE_CLIENT_HEADER_NAMES}
    for name in list(result):
        if name.lower() in names:
            value = result.pop(name)
            if _value(value):
                result[names[name.lower()]] = value
    for name, env_name, fallback in (
        ("User-Agent", "UPSTREAM_DEFAULT_USER_AGENT", "codex-cli"),
        ("Originator", "UPSTREAM_DEFAULT_ORIGINATOR", "codex_cli_rs"),
    ):
        result.setdefault(name, _value(os.environ.get(env_name, "")) or fallback)
    if provider == "opencode" and "X-Opencode-Session" not in result:
        for name in SESSION_HEADER_NAMES:
            value = result.get(CLIENT_HEADER_NAMES[name])
            if value:
                result["X-Opencode-Session"] = value
                break
    return result

File: worker/client-headers.mjs
// Shared client identity policy. Credentials and account headers are never inferred.
export const CLIENT_HEADER_NAMES = [
  "user-agent", "originator", "session-id", "thread-id", "session_id",
  "x-session-id", "x-codex-session-id", "x-session-affinity",
];
export const OPENCODE_CLIENT_HEADER_NAMES = [
  "x-opencode-session", "x-opencode-client", "x-opencode-project", "x-opencode-request",
];
const SESSION_HEADER_NAMES = [
  "thread-id", "session-id", "session_id", "x-session-id",
  "x-codex-session-id", "x-session-affinity",
];

function validValue(value) {
  return typeof value === "string" && !/[^\x20-\x7e]/.test(value) ? value.trim() : "";
}

export function clientContextHeaders(source, provider = "") {
  const input = new Headers(source);
  const names = provider === "opencode"
    ? [...CLIENT_HEADER_NAMES, ...OPENCODE_CLIENT_HEADER_NAMES]
    : CLIENT_HEADER_NAMES;
  const headers = new Headers();
  for (const name of names) {
    if (validValue(input.get(name))) headers.set(name, input.get(name));
  }
  return headers;
}

export function withClientDefaults(source, env = {}) {
  const headers = new Headers(source);
  for (const [name, setting, fallback] of [
    ["User-Agent", "UPSTREAM_DEFAULT_USER_AGENT", "codex-cli"],
    ["Originator", "UPSTREAM_DEFAULT_ORIGINATOR", "codex_cli_rs"],
  ]) {
    if (!validValue(headers.get(name))) {
      headers.set(name, validValue(env[setting]) || fallback);
    }
  }
  return headers;
}

export function withOpencodeSession(source, fallback = "") {
  const headers = new Headers(source);
  if (!validValue(headers.get("x-opencode-session"))) {
    headers.delete("x-opencode-session");
    const session = SESSION_HEADER_NAMES.map((name) => headers.get(name)).find(validValue);
    if (session || validValue(fallback)) headers.set("x-opencode-session", session || fallback);
  }
  return headers;
}
